### PR TITLE
Add test coverage for subjectrulesreview

### LINKED ISSUE #91042: Add test coverage for SelfSubjectRulesReview and node authorizer
An integration test should be added to confirm that the node authorizer interacts correctly with the `SelfSubjectRulesReview` API.

Test the 4 permutations of node authorizer=on/off with identity isNode=true/false.

xref: #90975 #91030

### PR DESCRIPTION
**Pull request by helenfeng737**
**What type of PR is this?**
 /kind bug
 /kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubernetes/issues/91042
Test coverage for change: https://github.com/kubernetes/kubernetes/pull/91030
**Special notes for your reviewer**:
Test passes with:
```
make test-integration WHAT=./test/integration/auth GOFLAGS="-v" KUBE_TEST_ARGS="-run ^TestSelfSubjectRulesReview$"
```
**Does this PR introduce a user-facing change?**:
```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
None

### CODE REVIEW

File: test/integration/auth/rulesreview_test.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (zjj2wry) [inline test/integration/auth/rulesreview_test.go:2]: 2020

File: test/integration/auth/rulesreview_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+	defer closeFn()
+
+	clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL})
+
+	tests := []struct {
+		name           string
+		username       string
+		srr            *authorizationapi.SelfSubjectRulesReview
+		expectedError  string
```
Comment (zjj2wry) [inline test/integration/auth/rulesreview_test.go:91]: seems expectedError not used

Comment (ZhiFeng1993) [reply test/integration/auth/rulesreview_test.go:91]: Good catch. Another use case added.

File: test/integration/auth/rulesreview_test.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	return authorizer.DecisionAllow, "", nil
+}
+
+func (r *srrAuthorizer) RulesFor(user user.Info, namespace string) ([]authorizer.ResourceRuleInfo, []authorizer.NonResourceRuleInfo, bool, error) {
+	if _, isNode := r.identifier.NodeIdentity(user); isNode {
+		// indicate nodes do not have fully enumerated permissions
+		return nil, nil, true, fmt.Errorf("node authorizer does not support user rule resolution")
+	}
+	return nil, nil, false, nil
+}
```
Comment (enj) [inline test/integration/auth/rulesreview_test.go:72]: None of this should be in the test.

File: test/integration/auth/rulesreview_test.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+func TestSelfSubjectRulesReviewNodeUser(t *testing.T) {
+	masterConfig := framework.NewIntegrationTestMasterConfig()
```
Comment (enj) [inline test/integration/auth/rulesreview_test.go:75]: Configure the node authorizer here.

File: test/integration/auth/rulesreview_test.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	masterConfig.GenericConfig.Authentication.Authenticator = authenticator.RequestFunc(func(req *http.Request) (*authenticator.Response, bool, error) {
+		return &authenticator.Response{
+			User: newNodeUserInfo(),
+		}, true, nil
+	})
+	masterConfig.GenericConfig.Authorization.Authorizer = srrAuthorizer{}
```
Comment (enj) [inline test/integration/auth/rulesreview_test.go:81]: Remove.

File: test/integration/auth/rulesreview_test.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+	defer closeFn()
+
+	clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL})
+
+	tests := []struct {
+		name           string
+		username       string
```
Comment (enj) [inline test/integration/auth/rulesreview_test.go:89]: Needs to be `user.Info` that needs to be impersonated.

File: test/integration/auth/rulesreview_test.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+		srr            *authorizationapi.SelfSubjectRulesReview
```
Comment (enj) [inline test/integration/auth/rulesreview_test.go:90]: Probably not needed as input to the test.

**Review by enj — CHANGES_REQUESTED** [review]
This does not actually test the node authorizer code.

It also `panic`s when run (but the test still passes):

```
E0621 22:07:36.138536   35067 wrap.go:39] apiserver panic'd on POST /apis/authorization.k8s.io/v1/selfsubjectrulesreviews
2020-06-21 22:07:36.138622 I | http: panic serving 127.0.0.1:65358: runtime error: invalid memory address or nil pointer dereference
```

```
1: running [Created by handlers.finishRequest @ rest.go:222]
    handlers               rest.go:233   finishRequest.func1.1(Duration(#7))
                           panic.go:967  panic(interface{}(0x4a9f0a0))
    selfsubjectrulesreview rest.go:75    (*REST).Create(*REST(0xc009ace5a0), Context(#3), Object(#2), ValidateObjectFunc(#1), *CreateOptions(0x0), 0x0, 0x0, ...)
    handlers               create.go:211 (*namedCreaterAdapter).Create(*namedCreaterAdapter(0xc003675540), Context(#3), string(0x0, len=0), Object(#2), ValidateObjectFunc(#1), *CreateOptions(0xc00b5a9db8), ...)
    handlers               create.go:145 createHandler.func1.1(NamedCreater(0x4b32e60), *RequestScope(0x0), Interface(0x0))
    handlers               create.go:169 createHandler.func1.2(NamedCreater(0x86f3ec0), *RequestScope(0xc00b5a9fc8), Interface(0x1340c43))
    handlers               rest.go:241   finishRequest.func1(#7, 0xc00be287e0, 0xc00be288a0, 0xc00be28840)
```

**Comment by enj** [comment]
I suggest using the `StartTestServer` function to start a test server (it has the node authorizer on by default and it is easy to turn it off via `ModifyServerRunOptions`).  Use impersonation to "create" an identity that matches a node.

File: test/integration/auth/rulesreview_test.go [hunk 8]
```diff
@@ -0,0 +1,206 @@
+/*
```
Comment (enj) [inline test/integration/auth/rulesreview_test.go:1]: Call this file `node_rule_resolution_test.go`.

Comment (ZhiFeng1993) [reply test/integration/auth/rulesreview_test.go:1]: Ok.

File: test/integration/auth/rulesreview_test.go (hunk 4 above)
Comment (ZhiFeng1993) [reply test/integration/auth/rulesreview_test.go:75]: ok

File: test/integration/auth/rulesreview_test.go (hunk 5 above)
Comment (ZhiFeng1993) [reply test/integration/auth/rulesreview_test.go]: updated with starttestserver as suggested.

File: test/integration/auth/rulesreview_test.go (hunk 6 above)
Comment (ZhiFeng1993) [reply test/integration/auth/rulesreview_test.go]: updated as suggested

**Comment by ZhiFeng1993** [comment]
@enj 
Thanks for the advice. I have made big changes to this PR, please take a look again. 
Basically changes are: 
1. Using startTestServer, which enables node authorizer by default to test subjectrulesviews api.
2. Impersonated two types of clients 'system:node:foo' and 'user:foo', which will hit two paths in node authorizer's rule resolver function. 
3. Changed the file name to be node_authorizer_rule_resolution as suggested and test names accordingly.

**Pull request closed without merging**